File: frontend-web/src/lib/passwordPolicy.ts
export const PASSWORD_MIN_LENGTH = 14;

export type PasswordViolation =
  | "min_length"
  | "uppercase"
  | "lowercase"
  | "number"
  | "special_char"
  | "contains_email"
  | "contains_name"
  | "denylist_common";

const COMMON_WEAK_PASSWORDS = new Set([
  "password",
  "password123",
  "123456",
  "123456789",
  "12345678",
  "qwerty",
  "abc123",
  "letmein",
  "111111",
  "iloveyou",
  "admin",
  "welcome",
  "monkey",
  "dragon",
  "football",
  "baseball",
  "123123",
  "qwerty123",
  "zaq12wsx",
  "trustno1",
  "passw0rd",
  "sunshine",
  "princess",
  "login",
  "whatever",
]);

function normalize(value?: string | null) {
  return (value ?? "").trim().toLowerCase();
}

function hasUppercase(value: string) {
  return /[A-Z]/.test(value);
}

function hasLowercase(value: string) {
  return /[a-z]/.test(value);
}

function hasNumber(value: string) {
  return /[0-9]/.test(value);
}

function hasSpecial(value: string) {
  return /[^A-Za-z0-9]/.test(value);
}

export function evaluatePassword(
  password: string,
  {
    email,
    name,
    minLength = PASSWORD_MIN_LENGTH,
  }: { email?: string | null; name?: string | null; minLength?: number } = {}
): PasswordViolation[] {
  const pw = password ?? "";
  const violations: PasswordViolation[] = [];
  const normalizedPassword = pw.toLowerCase();

  if ((pw ?? "").length < Math.max(minLength, 1)) violations.push("min_length");
  if (!hasUppercase(pw)) violations.push("uppercase");
  if (!hasLowercase(pw)) violations.push("lowercase");
  if (!hasNumber(pw)) violations.push("number");
  if (!hasSpecial(pw)) violations.push("special_char");

  const normalizedEmail = normalize(email);
  if (normalizedEmail) {
    if (normalizedPassword.includes(normalizedEmail)) {
      violations.push("contains_email");
    } else {
      const local = normalizedEmail.split("@")[0];
      if (local && normalizedPassword.includes(local)) violations.push("contains_email");
    }
  }

  const normalizedName = normalize(name);
  if (normalizedName && normalizedPassword.includes(normalizedName)) {
    violations.push("contains_name");
  }

  if (COMMON_WEAK_PASSWORDS.has(normalizedPassword)) {
    violations.push("denylist_common");
  }

  const unique: PasswordViolation[] = [];
  const seen = new Set<string>();
  for (const v of violations) {
    if (seen.has(v)) continue;
    seen.add(v);
    unique.push(v);
  }
  return unique;
}

export function describeViolation(code: PasswordViolation, { minLength = PASSWORD_MIN_LENGTH }: { minLength?: number } = {}): string {
  switch (code) {
    case "min_length":
      return `At least ${minLength} characters`;
    case "uppercase":
      return "At least one uppercase letter (A-Z)";
    case "lowercase":
      return "At least one lowercase letter (a-z)";
    case "number":
      return "At least one number (0-9)";
    case "special_char":
      return "At least one special character (!@#$, etc.)";
    case "contains_email":
      return "Cannot contain your email address";
    case "contains_name":
      return "Cannot contain your name";
    case "denylist_common":
      return "Cannot be a common password";
    default:
      return "Password requirement not met";
  }
}

File: backend/app/core/password_policy.py
from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING, List

from fastapi import HTTPException, status

from app.core.config import settings

if TYPE_CHECKING:
    from app.models.user import User

COMMON_WEAK_PASSWORDS = {
    "password",
    "password123",
    "123456",
    "123456789",
    "12345678",
    "qwerty",
    "abc123",
    "letmein",
    "111111",
    "iloveyou",
    "admin",
    "welcome",
    "monkey",
    "dragon",
    "football",
    "baseball",
    "123123",
    "qwerty123",
    "zaq12wsx",
    "trustno1",
    "passw0rd",
    "sunshine",
    "princess",
    "login",
    "whatever",
}

_UPPERCASE_RE = re.compile(r"[A-Z]")
_LOWERCASE_RE = re.compile(r"[a-z]")
_NUMBER_RE = re.compile(r"[0-9]")
_SPECIAL_RE = re.compile(r"[^A-Za-z0-9]")


def _normalize(value: str | None) -> str:
    return (value or "").strip().lower()


def evaluate_password(
    password: str,
    *,
    email: str | None = None,
    username: str | None = None,
) -> List[str]:
    """
    Returns a list of violation codes if the password does not meet policy.
    """
    pw = password or ""
    violations: list[str] = []
    min_length = max(int(getattr(settings, "PASSWORD_MIN_LENGTH", 14) or 0), 1)

    if len(pw) < min_length:
        violations.append("min_length")
    if not _UPPERCASE_RE.search(pw):
        violations.append("uppercase")
    if not _LOWERCASE_RE.search(pw):
        violations.append("lowercase")
    if not _NUMBER_RE.search(pw):
        violations.append("number")
    if not _SPECIAL_RE.search(pw):
        violations.append("special_char")

    normalized_pw = pw.lower()

    email_norm = _normalize(email)
    if email_norm and email_norm in normalized_pw:
        violations.append("contains_email")
    else:
        local_part = email_norm.split("@")[0] if email_norm else ""
        if local_part and local_part in normalized_pw:
            violations.append("contains_email")

    username_norm = _normalize(username)
    if username_norm and username_norm in normalized_pw:
        violations.append("contains_name")

    if normalized_pw in COMMON_WEAK_PASSWORDS:
        violations.append("denylist_common")

    # Remove duplicates while preserving original order.
    seen = set()
    result: list[str] = []
    for v in violations:
        if v in seen:
            continue
        seen.add(v)
        result.append(v)
    return result


def ensure_strong_password(password: str, *, email: str | None = None, username: str | None = None) -> None:
    violations = evaluate_password(password, email=email, username=username)
    if violations:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "message": "Password does not meet requirements.",
                "details": {"code": "WEAK_PASSWORD", "violations": violations},
            },
        )

